kbase-extension/static/kbase/js/common/cellComponents/paramsWidget.js

```diff
@@ -0,0 +1,551 @@
+                                try {
+                                    return makeFieldWidget(appSpec, spec, initialParams[spec.id])
+                                        .then(function (widget) {
+                                            widgets.push(widget);
+
+                                            return widget.start({
+                                                node: document.getElementById(parameterParams.view[spec.id].id)
+                                            });
+                                        });
+                                } catch (ex) {
+                                    console.error('Error making input field widget', ex);
+                                    var errorDisplay = div({ style: { border: '1px red solid' } }, [
+                                        ex.message
+                                    ]);
+                                    document.getElementById(parameterParams.view[spec.id].id).innerHTML = errorDisplay;
+                                }
+                            }));
+                        }
+                    })
+                    .then(function () {
+                        renderAdvanced();
+                    });
+            });
+        }
+
+        function start(arg) {
+            return Promise.try(function () {
+                // send parent the ready message
+
+                doAttach(arg.node);
+
+                model.setItem('appSpec', arg.appSpec);
+                model.setItem('parameters', arg.parameters);
+
+                paramsBus.on('parameter-changed', function (message) {
+                    // Also, tell each of our inputs that a param has changed.
+                    // TODO: use the new key address and subscription
+                    // mechanism to make this more efficient.
+                    widgets.forEach(function (widget) {
+                        widget.bus.send(message, {
+                            key: {
+                                type: 'parameter-changed',
+                                parameter: message.parameter
+                            }
+                        });
+                        // bus.emit('parameter-changed', message);
+                    });
+                });
+
+
+                return renderParameters()
+                    .then(function () {
+                        // do something after success
+                        attachEvents();
+                    })
+                    .catch(function (err) {
+                        // do somethig with the error.
+                        console.error('ERROR in start', err);
+                    });
+            });
+
+            
+        }
+
+        function stop() {
+            return Promise.try(function () {
+                // really unhook things here.
+            });
+        }
+
+        // CONSTRUCTION
+
+        bus = runtime.bus().makeChannelBus({ description: 'A app params widget' });
```

Comment:
Can this be moved to the top of the file where the rest of variables are defined, just for consistency?